test/gtest/ucs/test_conn_match.cc

```diff
@@ -0,0 +1,251 @@
+
+    address_generators.push_back(new uuid_generator);
+    address_generators.push_back(new string_generator);
+
+    for (std::vector<address_generator*>::const_iterator it =
+             address_generators.begin(); it != address_generators.end(); it++) {
+        address_generator *generator = *it;
+        std::vector<std::vector<conn_elem_t> > elems(max_addresses);
+
+        conn_match_init(generator->length());
+
+        for (size_t i = 0; i < max_addresses; i++) {
+            const void *dest_address = generator->generate();
+            ucs_conn_sn_t num_conns  = ucs::rand() % max_conns + 1;
+            elems[i].resize(num_conns);
+
+            for (ucs_conn_sn_t conn = 0; conn < elems[i].size(); conn++) {
+                conn_elem_t *conn_elem = &elems[i][conn];
+                EXPECT_EQ(conn, get_next_sn(dest_address));
+
+                conn_elem->dest_address = new uint8_t[generator->length()];
+                ASSERT_TRUE(conn_elem->dest_address != NULL);
+                memcpy(conn_elem->dest_address, dest_address,
+                       generator->length());
+
+                ucs_conn_match_queue_type_t queue_type = (ucs::rand() & 1) ?
+                                                         UCS_CONN_MATCH_QUEUE_EXP :
+                                                         UCS_CONN_MATCH_QUEUE_UNEXP;
+                insert(dest_address, conn, queue_type, *conn_elem);
+                EXPECT_EQ(queue_type, conn_elem->queue_type);
+                EXPECT_EQ(conn, conn_elem->conn_sn);
+            }
+        }
+
+        for (size_t i = 0; i < max_addresses; i++) {
+            for (ucs_conn_sn_t conn = 0; conn < elems[i].size(); conn++) {
+                conn_elem_t *conn_elem   = &elems[i][conn];
+                const void *dest_address = conn_elem->dest_address;
+                conn_elem_t *test_conn_elem;
+                ucs_conn_match_queue_type_t another_queue_type =
+                    (conn_elem->queue_type == UCS_CONN_MATCH_QUEUE_EXP) ?
+                    UCS_CONN_MATCH_QUEUE_UNEXP : UCS_CONN_MATCH_QUEUE_EXP;
+
+                /* must not find this element in the another queue */
+                test_conn_elem = retrieve(dest_address, conn_elem->conn_sn,
+                                          another_queue_type);
+                EXPECT_EQ(NULL, test_conn_elem);
+
+                test_conn_elem = retrieve(dest_address, conn_elem->conn_sn,
+                                          conn_elem->queue_type);
+                EXPECT_EQ(conn_elem, test_conn_elem);
+
+                /* subsequent retrieving of the same connection element must return NULL */
+                test_conn_elem = retrieve(dest_address, conn_elem->conn_sn,
+                                          conn_elem->queue_type);
+                EXPECT_EQ(NULL, test_conn_elem);
+
+                insert(dest_address, conn_elem->conn_sn, another_queue_type,
+                       *conn_elem);
```

Comment:
how come we are able to insert to another queue with the same address,sn?
seems like an unexpected usage of the data structure - after we retrieved sam connection cannot arrive again?

Reply:
why not? e.g. in TCP can be some "ghost" connection that will be added to an unexpected queue after connection created thru API is closed by the user. And then this "ghost" connection will be closed since the error will be detected